homeassistant/components/onvif/device.py

```diff
@@ -0,0 +1,401 @@
+
+                LOGGER.debug("TimeZone for date/time: %s", tzone)
+
+                LOGGER.debug("Camera date/time: %s", cam_date)
+
+                LOGGER.debug("Camera date/time in UTC: %s", cam_date_utc)
+
+                LOGGER.debug("System date/time: %s", system_date)
+
+                dt_diff = cam_date - system_date
+                dt_diff_seconds = dt_diff.total_seconds()
+
+                if dt_diff_seconds > 5:
+                    LOGGER.warning(
+                        "The date/time on the camera (UTC) is '%s', "
+                        "which is different from the system '%s', "
+                        "this could lead to authentication issues",
+                        cam_date_utc,
+                        system_date,
+                    )
+        except ServerDisconnectedError as err:
+            LOGGER.warning(
+                "Couldn't get camera '%s' date/time. Error: %s", self.name, err
+            )
+
+    async def async_get_device_info(self) -> DeviceInfo:
+        """Obtain information about this device."""
+        devicemgmt = self.device.create_devicemgmt_service()
+        device_info = await devicemgmt.GetDeviceInformation()
+        return DeviceInfo(
+            device_info.Manufacturer,
+            device_info.Model,
+            device_info.FirmwareVersion,
+            self.config_entry.unique_id,
+        )
+
+    async def async_get_capabilities(self):
+        """Obtain information about the available services on the device."""
+        media_service = self.device.create_media_service()
+        capabilities = await media_service.GetServiceCapabilities()
+        ptz = False
+        try:
+            self.device.get_definition("ptz")
+            ptz = True
+        except ONVIFError:
+            pass
+        return Capabilities(capabilities.SnapshotUri, ptz)
+
+    async def async_get_profiles(self) -> List[Profile]:
+        """Obtain media profiles for this device."""
+        media_service = self.device.create_media_service()
+        result = await media_service.GetProfiles()
+        profiles = [
+            Profile(
+                key,
+                profile.token,
+                profile.Name,
+                Video(
+                    profile.VideoEncoderConfiguration.Encoding,
+                    Resolution(
+                        profile.VideoEncoderConfiguration.Resolution.Width,
+                        profile.VideoEncoderConfiguration.Resolution.Height,
+                    ),
+                ),
+                PTZ(
+                    profile.PTZConfiguration.DefaultContinuousPanTiltVelocitySpace
+                    is not None,
+                    profile.PTZConfiguration.DefaultRelativePanTiltTranslationSpace
+                    is not None,
+                    profile.PTZConfiguration.DefaultAbsolutePantTiltPositionSpace
+                    is not None,
+                )
+                if profile.PTZConfiguration
+                else None,
```

Comment:
I don't think that this code is very readable. I think it would benefit from not using a list comprehension and not using a multiline inline if-statements.